presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergFileWriter.java

```diff
@@ -0,0 +1,177 @@
+                ImmutableMap.of(
+                        "iceberg.parquet.metadata-cache-enabled", "true",
+                        "iceberg.parquet.metadata-cache-size", "100MB",
+                        "iceberg.parquet.metadata-cache-ttl-since-last-access", "1h"),
+                PARQUET,
+                false,
+                true,
+                OptionalInt.of(2),
+                Optional.empty());
+    }
+
+    @BeforeClass
+    public void setup()
+    {
+        ConnectorContext context = new TestingConnectorContext();
+        String catalogName = "testing_catalog";
+        Bootstrap app = new Bootstrap(
+                new EventModule(),
+                new MBeanModule(),
+                new JsonModule(),
+                new IcebergCommonModule(catalogName),
+                new IcebergCatalogModule(catalogName, Optional.of(new InMemoryCachingHiveMetastore(
+                        new FileHiveMetastore(
+                                getHdfsEnvironment(),
+                                "test_directory",
+                                "test_user"),
+                        new ThreadPoolExecutor(1, 1, 10000, TimeUnit.MILLISECONDS, new BlockingArrayQueue<>()),
+                        new HiveMetastoreCacheStats(),
+                        new MetastoreClientConfig()
+                ))),
+                new HiveS3Module(catalogName),
+                new HiveGcsModule(),
+                new HiveAuthenticationModule(),
+                new CachingModule(),
+                new HiveCommonModule(),
+                binder -> {
+                    MBeanServer platformMBeanServer = ManagementFactory.getPlatformMBeanServer();
+                    binder.bind(MBeanServer.class).toInstance(new RebindSafeMBeanServer(platformMBeanServer));
+                    binder.bind(NodeVersion.class).toInstance(new NodeVersion(context.getNodeManager().getCurrentNode().getVersion()));
+                    binder.bind(NodeManager.class).toInstance(context.getNodeManager());
+                    binder.bind(TypeManager.class).toInstance(context.getTypeManager());
+                    binder.bind(PageIndexerFactory.class).toInstance(context.getPageIndexerFactory());
+                    binder.bind(StandardFunctionResolution.class).toInstance(context.getStandardFunctionResolution());
+                    binder.bind(FunctionMetadataManager.class).toInstance(context.getFunctionMetadataManager());
+                    binder.bind(RowExpressionService.class).toInstance(context.getRowExpressionService());
+                    binder.bind(FilterStatsCalculatorService.class).toInstance(context.getFilterStatsCalculatorService());
+                });
+
+        Injector injector = app
+                .doNotInitializeLogging()
+                .setRequiredConfigurationProperties(ImmutableMap.of())
+                .initialize();
+
+        this.icebergFileWriterFactory = injector.getInstance(IcebergFileWriterFactory.class);
```

Comment:
I've been trying to figure out the best way to instantiate this class in the tests here but it has been a very convoluted process. Right now I have this injector implementation but it is still giving me some explicit bindings error for my hive metastore. Does anybody else know if there is a better way to go about this?

Reply:
Maybe we can build `IcebergFileWriterFactory` directly. I think the following code could work, you can have a try:

```
        Session session = getQueryRunner().getDefaultSession();
        ConnectorSession connectorSession = session.toConnectorSession(new ConnectorId("iceberg"));
        TypeManager typeManager = new TypeManager()
        {
            @Override
            public Type getType(TypeSignature signature)
            {
                for (Type type : getTypes()) {
                    if (signature.getBase().equals(type.getTypeSignature().getBase())) {
                        return type;
                    }
                }
                return null;
            }

            @Override
            public Type getParameterizedType(String baseTypeName, List<TypeSignatureParameter> typeParameters)
            {
                return getType(new TypeSignature(baseTypeName, typeParameters));
            }

            @Override
            public boolean canCoerce(Type actualType, Type expectedType)
            {
                throw new UnsupportedOperationException();
            }

            private List<Type> getTypes()
            {
                return ImmutableList.of(BOOLEAN, INTEGER, BIGINT, DOUBLE, VARCHAR, VARBINARY, TIMESTAMP, DATE, HYPER_LOG_LOG);
            }
        };
        HdfsContext hdfsContext = new HdfsContext(connectorSession);
        HdfsEnvironment hdfsEnvironment = IcebergDistributedTestBase.getHdfsEnvironment();
        IcebergFileWriterFactory icebergFileWriterFactory = new IcebergFileWriterFactory(hdfsEnvironment, typeManager,
                new FileFormatDataSourceStats(), new NodeVersion("test"), new OrcFileWriterConfig(), HiveDwrfEncryptionProvider.NO_ENCRYPTION);

        Path path = new Path(createTempDir().getAbsolutePath());
        Schema schema = toIcebergSchema(ImmutableList.of(new ColumnMetadata("a", VARCHAR), new ColumnMetadata("b", INTEGER), new ColumnMetadata("c", TIMESTAMP)));
        IcebergFileWriter icebergFileWriter = icebergFileWriterFactory.createFileWriter(path, schema, new JobConf(), connectorSession,
                hdfsContext, FileFormat.PARQUET, MetricsConfig.getDefault());

        List<Page> input = rowPagesBuilder(VARCHAR, BIGINT, TIMESTAMP)
                .addSequencePage(100, 0, 0, 123)
                .addSequencePage(100, 100, 100, 223)
                .addSequencePage(100, 200, 200, 323)
                .build();
        for (Page page : input) {
            icebergFileWriter.appendRows(page);
        }
        icebergFileWriter.commit();
```

Reply:
Thanks! I was able to add this into the test. It is currently failing but I'm pretty sure this is due to the reader not being fixed yet. Will work on that this afternoon